drivers: mbox: add initial driver for nRF VEVIF (VPR event)

jaz1-nordic opened the pull request.

jaz1-nordic committed:
drivers: mbox: add initial driver for nRF VEVIF event

Add a mailbox driver for VEVIF events (VPR irq).
The driver can be built in either 'rx' or 'tx' configuration.
The VPR sends the event, so it uses the 'tx' configuration,
while the master core uses the 'rx' configuration of the driver
to receive the VPR events.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
dts: nordic: nrf54l15: Add mbox VEVIF nodes

Add a mbox VEVIF nodes to be used for communicating FLPR -> APP.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
samples: mbox: Add nRF54L15 to MBOX sample targets

Add nRF54L15 APP and FLPR cores to ping-pong sample application.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

hubertmis commented on drivers/mbox/mbox_nrf_vevif_event_local.c:
```diff
@@ -0,0 +1,147 @@
+
+	for (uint8_t id = 0; id < NRF_VPR_EVENTS_TRIGGERED_COUNT; id++) {
+		nrf_vpr_event_t event = nrf_vpr_triggered_event_get(id);
+		if ((cbs->enabled_mask & BIT(id)) && nrf_vpr_event_check(config->vpr, event)) {
```
I'm curious if this condition can lead to some kind of a IRQ loop.

Let's assume there is an interrupt enabled for event X by `nrf_vpr_int_enable(config->vpr, BIT(X));`, but bit X is cleared in `cbs->enabled_mask`.
1. Event X is triggered
2. Event X causes IRQ in ARM
3. ARM handles IRQ with `vevif_event_local_isr()`
4. ARM iterates over all events, but this condition evaluates to `false` for all iterations
5. ARM exits the loop and the IRQ handler
6. The event is still pending, so it causes IRQ in ARM (go to step 2)

hubertmis replied:
On the other hand, the scenario in which my assumption is present is invalid. So maybe we can let the watchdog handle the problem

hubertmis commented on drivers/mbox/CMakeLists.txt:
```diff
@@ -14,3 +14,5 @@ zephyr_library_sources_ifdef(CONFIG_MBOX_NRF_VEVIF_LOCAL mbox_nrf_vevif_local.c)
 zephyr_library_sources_ifdef(CONFIG_MBOX_NRF_VEVIF_REMOTE mbox_nrf_vevif_remote.c)
 zephyr_library_sources_ifdef(CONFIG_MBOX_NRF_BELLBOARD_LOCAL mbox_nrf_bellboard_local.c)
 zephyr_library_sources_ifdef(CONFIG_MBOX_NRF_BELLBOARD_REMOTE mbox_nrf_bellboard_remote.c)
+zephyr_library_sources_ifdef(CONFIG_MBOX_NRF_VEVIF_EVENT_LOCAL mbox_nrf_vevif_event_local.c)
+zephyr_library_sources_ifdef(CONFIG_MBOX_NRF_VEVIF_EVENT_REMOTE mbox_nrf_vevif_event_remote.c)
```
I'm thinking if it would be clearer to rename all local/remote labels for VEVIF and BELLBOARD to something more descriptive like "sender/receiver", "source/destination", "trigger/recipient"

jaz1-nordic committed:
drivers: mbox: nrf: Change VEVIFs and BELLBOARD nomenclature

Renaming 'LOCAL' to 'RX' and 'REMOTE' to 'TX'.
This seems more descriptive and intuitive to use.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
dts: nordic: Align boards dts to new VEVIF, BELLBOARD nomenclature

After changing the VEVIF and BELLBOARD names,
the dts for the individual boards must be aligned.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

masz-nordic commented on drivers/mbox/mbox_nrf_vevif_event_rx.c:
```diff
@@ -0,0 +1,152 @@
+#define EVENTS_IDX_MIN 11U
+#define EVENTS_IDX_MAX 17U
+#else
+#define EVENTS_IDX_MIN VPR_EVENTS_TRIGGERED_MinIndex
+#define EVENTS_IDX_MAX VPR_EVENTS_TRIGGERED_MaxIndex
```
```suggestion
#define EVENTS_IDX_MIN NRF_VPR_EVENTS_TRIGGERED_MIN
#define EVENTS_IDX_MAX NRF_VPR_EVENTS_TRIGGERED_MAX
```

masz-nordic commented on drivers/mbox/mbox_nrf_vevif_event_tx.c:
```diff
@@ -0,0 +1,81 @@
+
+#if defined(CONFIG_SOC_NRF54L15_ENGA_CPUFLPR)
+#define EVENTS_IDX_MAX 17U
+#else
+#define EVENTS_IDX_MAX VPR_EVENTS_TRIGGERED_MaxIndex
```
```suggestion
#else
#define EVENTS_IDX_MAX NRF_VPR_EVENTS_TRIGGERED_MAX
```

masz-nordic commented on drivers/mbox/mbox_nrf_vevif_event_tx.c:
```diff
@@ -0,0 +1,81 @@
+};
+
+#define VEVIF_EVENT_REMOTE_DEFINE(inst)                                                            \
+	BUILD_ASSERT(DT_INST_PROP(inst, nordic_events) <= VPR_EVENTS_TRIGGERED_MaxCount,           \
```
```suggestion
	BUILD_ASSERT(DT_INST_PROP(inst, nordic_events) <= NRF_VPR_EVENTS_TRIGGERED_COUNT,           \
```

masz-nordic commented on drivers/mbox/mbox_nrf_vevif_task_rx.c:
```diff
@@ -0,0 +1,133 @@
+#define TASKS_IDX_MIN 11U
+#define TASKS_IDX_MAX 17U
+#else
+#define TASKS_IDX_MIN VPR_TASKS_TRIGGER_MinIndex
+#define TASKS_IDX_MAX VPR_TASKS_TRIGGER_MaxIndex
```
```suggestion
#define TASKS_IDX_MIN NRF_VPR_TASKS_TRIGGER_MIN
#define TASKS_IDX_MAX NRF_VPR_TASKS_TRIGGER_MAX
```

masz-nordic commented on drivers/mbox/mbox_nrf_vevif_task_tx.c:
```diff
@@ -3,31 +3,37 @@
+#if defined(CONFIG_SOC_NRF54L15_ENGA_CPUAPP)
+#define TASKS_IDX_MAX 17U
+#else
+#define TASKS_IDX_MAX VPR_TASKS_TRIGGER_MaxIndex
```
```suggestion
#define TASKS_IDX_MAX NRF_VPR_TASKS_TRIGGER_MAX
```

jaz1-nordic commented:
@anangl please take a look. PR was already approved by @carlocaione and @hubertmis.

gmarull commented on drivers/mbox/mbox_nrf_vevif_event_tx.c:
```diff
@@ -0,0 +1,80 @@
+	DEVICE_DT_INST_DEFINE(inst, NULL, NULL, NULL, &conf##inst, POST_KERNEL,                    \
+			      CONFIG_MBOX_INIT_PRIORITY, &vevif_event_tx_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(VEVIF_EVENT_REMOTE_DEFINE)
```
is multi-instance really needed here? This is using CSR instructions... Please, check https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mbox/mbox_nrf_vevif_local.c

jaz1-nordic replied:
You're right, reworked to a single-instance driver.

gmarull commented on dts/bindings/mbox/nordic,nrf-vevif-event-rx.yaml:
```diff
@@ -0,0 +1,47 @@
+        interrupts = <76 NRF_DEFAULT_IRQ_PRIORITY>;
+        #mbox-cells = <1>;
+        nordic,events = <1>;
+        nordic,events-mask: <0x00008000>;
```
wrong syntax, please fix

jaz1-nordic replied:
The same error was in yaml for vevif_task, fixed.

gmarull commented on drivers/mbox/CMakeLists.txt:
```diff
@@ -12,6 +12,8 @@ zephyr_library_sources_ifdef(CONFIG_MBOX_NXP_MAILBOX   mbox_nxp_mailbox.c)
 zephyr_library_sources_ifdef(CONFIG_MBOX_ANDES_PLIC_SW   mbox_andes_plic_sw.c)
```
commit description for bb585c08cc5f0040a441d1bc5a0eafafafc01b16 is misleading, existence of `interrupts` property is irrelevant

jaz1-nordic replied:
Description changed.

gmarull reviewed (changes requested):
looks like code can be simplified and aligned with 54h20 vevif

jaz1-nordic commented on drivers/mbox/mbox_nrf_vevif_event_rx.c:
```diff
@@ -0,0 +1,152 @@
+
+#include <hal/nrf_vpr.h>
+
+#if defined(CONFIG_SOC_NRF54L15_ENGA_CPUAPP)
+#define EVENTS_IDX_MIN 11U
+#define EVENTS_IDX_MAX 17U
+#else
+#define EVENTS_IDX_MIN NRF_VPR_EVENTS_TRIGGERED_MIN
+#define EVENTS_IDX_MAX NRF_VPR_EVENTS_TRIGGERED_MAX
+#endif
```
No, it's because MDK has incorrect values.

gmarull reviewed (commented):
At a first glance, this looks fairly similar to what we have for nRF54H20, however, different terms are used (e.g. tx/rx vs local/remote, etc). How different is this VEVIF IP in 54L15? My feeling is that there should be more alignment to have consistency. I can read the spec to compare and make more detailed suggestions (if they really apply)

masz-nordic commented:
It's functionally the same IP, but on L15 there is no bellboard so you need to use VEVIF events to trigger IRQ on ARM.
What you refer to as H20 stuff was VEVIF tasks (used for ARM->VPR path), which are now aligned as well to rx/tx naming.

magp-nordic commented on drivers/mbox/mbox_nrf_vevif_task_tx.c:
```diff
@@ -3,31 +3,37 @@
  */
 
-#define DT_DRV_COMPAT nordic_nrf_vevif_remote
+#define DT_DRV_COMPAT nordic_nrf_vevif_task_tx
```
Shouldn't it be rx? If I understand correctly, bellboard is used to trigger IRQ on other core from VPR, so "remote" is equivalent to tx, but vevif task is used to trigger IRQ on VPR from other core, so, in this case, "remote" equivalent should be rx. Or am I missing something?

jaz1-nordic replied:
The earlier nomenclature may have been confusing.

Currently we have `tx` and `rx` depending on the purpose of a specific node.
If we use `vevif events`, the signal goes from the VPR core to the remote core.
So, we have the `vevif_event_tx` node, which we set up in the VPR dts and as the name suggests, we use to transmit the signals (tx) from the VPR core to the remote core.
On the other hand, we have the `vevif_event_rx` node, which we configure in the remote core dts and use to receive the signals (rx) on the remote core.

Similarly, we have `vevif tasks` that are used for signaling from the remote core to the VPR core.
So, we have the `vevif_task_rx` node, which we set up in the VPR dts and as the name suggests, we use to receive the signals (rx) on the VPR core.
On the other hand, we have the `vevif_task_tx` node, which we configure in the remote core dts and use to transmit the signals (tx) from the remote core to the VPR core.
It seems quite analogous to me.

magp-nordic replied:
Ok, I understand now, I thought that "remote" always refers to RISCV core and "local" always refers to other core (e.g. ARM). I did not know that "remote" means core that triggers IRQs.

hubertmis commented on drivers/mbox/mbox_nrf_vevif_event_rx.c:
```diff
@@ -0,0 +1,152 @@
+			return -EALREADY;
+		}
+
+		cbs->enabled_mask |= BIT(id);
```
In general, reading register values is slooooow. If we can keep a copy in RAM it's much better for performance. Not sure if it was the reasoning here though.

hubertmis replied:
note: there are exception to this rule. Eg VPR accessing its internal peripherals or peripherals attached to its closest memory bus is extremely fast.

jaz1-nordic commented on drivers/mbox/mbox_nrf_vevif_event_rx.c:
```diff
@@ -0,0 +1,152 @@
+                                                                                                   \
+	static void irq_connect##inst(void)                                                        \
+	{                                                                                          \
+		IRQ_CONNECT(DT_IRQN(DT_DRV_INST(inst)), DT_IRQ(DT_DRV_INST(inst), priority),       \
+			    vevif_event_rx_isr, (const void *)DEVICE_DT_GET(DT_DRV_INST(inst)),    \
+			    0);                                                                    \
+		irq_enable(DT_IRQN(DT_DRV_INST(inst)));                                            \
```
Such that the driver is multi-instance and interrupts must be correctly connected and enabled for the appropriate instance.
At the initialization level, we are not able to correctly determine for which instance we are currently calling init, so we need to create the irq_connect functions at the build stage. The same applies to data and conf structures.

carlescufi commented:
@gmarull PTAL this week

gmarull reviewed (commented):
this code continues to confuse me, it's essentially a copy of what we have for vevif local/remote but using events (not tasks), and using different terminology, rx/tx vs local/remote (this code still has eg `VEVIF_EVENT_LOCAL_DEFINE` indicating it's been based on what we already have). We should align, and potentially re-use driver code.

masz-nordic commented:
>this code still has eg VEVIF_EVENT_LOCAL_DEFINE indicating it's been based on what we already have

@jaz1-nordic can you address this?

>We should align, and potentially re-use driver code.

That is planned, but out of scope of this PR.

gmarull commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -44,7 +44,7 @@
 			nordic,bus-width = <32>;
 
 			cpuppr_vevif_local: mailbox {
```
this still keeps `_local` nomenclature, should be updated to `_rx`

gmarull commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -413,7 +413,7 @@
 			};
 
 			cpusys_vevif_remote: mailbox@8c8000 {
```
same, here with `_tx`

gmarull commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -518,7 +518,7 @@
 				ranges = <0x0 0x908000 0x4000>;
 
 				cpuppr_vevif_remote: mailbox@0 {
```
same

gmarull commented on samples/drivers/mbox/boards/nrf54l15pdk_nrf54l15_cpuapp.overlay:
```diff
@@ -0,0 +1,20 @@
+/ {
+	mbox-consumer {
+		compatible = "vnd,mbox-consumer";
+		mboxes = <&mbox_cpuapp_rx_cpuflpr 15>, <&mbox_cpuapp_tx_cpuflpr 16>;
```
note: nodelabels should likely be `cpuapp_vevif_rx` and `cpuapp_vevif_tx` (no mbox, maybe not remote core as there's one only)

jaz1-nordic committed:
samples: ipc: icmsg: Align icmsg sample for nRF54L15

Add nRF54L15 APP and FLPR cores to icmsg sample application.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

nashif commented:
needs rebase

carlescufi commented:
> needs rebase

@nashif  rebased

kartben commented on dts/arm/nordic/nrf54l15_cpuapp.dtsi:
```diff
@@ -37,6 +36,27 @@ cpuflpr_vevif: &cpuflpr_vevif_remote {};
+&cpuflpr_vpr {
+	cpuapp_vevif_rx: mailbox@1 {
+		compatible = "nordic,nrf-vevif-event-rx";
+		reg = <0x0 0x1000>;
```
```
unit address and first address in 'reg' (0x5004c000) don't match for /soc/peripheral@50000000/vpr@4c000/mailbox@1
```

@jaz1-nordic I think this should be `reg = <0x1 0x1000>;`?

kartben replied:
cc @gmarull @anangl

gmarull replied:
needs fixing @jaz1-nordic
